File: Fraudulent-Activity-Notifications.cpp
#include <bits/stdc++.h>

using namespace std;

string ltrim(const string &);
string rtrim(const string &);
vector<string> split(const string &);

/*
 * Complete the 'activityNotifications' function below.
 *
 * The function is expected to return an INTEGER.
 * The function accepts following parameters:
 *  1. INTEGER_ARRAY expenditure
 *  2. INTEGER d
 */

void insertInOrder(std::vector<int>& arr, int x) {
    auto pos = std::upper_bound(arr.begin(), arr.end(), x);
    arr.insert(pos, x);
}

void removeFromOrdered(std::vector<int>& arr, int x) {
    auto pos = std::lower_bound(arr.begin(), arr.end(), x);
    arr.erase(pos);
}

int activityNotifications(std::vector<int> expenditure, int d) {
    bool even = (d % 2 == 0);
    int notifications = 0;

    std::vector<int> ordered(expenditure.begin(), expenditure.begin() + d);
    std::sort(ordered.begin(), ordered.end());

    for (int i = d; i < expenditure.size(); i++) {
        double median = even ? (ordered[d/2 - 1] + ordered[d/2]) / 2.0 : ordered[d / 2];

        if (2 * median <= expenditure[i]) {
            notifications++;
        }

        removeFromOrdered(ordered, expenditure[i - d]);
        insertInOrder(ordered, expenditure[i]);
    }

    return notifications;
}

int main()
{
    ofstream fout(getenv("OUTPUT_PATH"));

    string first_multiple_input_temp;
    getline(cin, first_multiple_input_temp);

    vector<string> first_multiple_input = split(rtrim(first_multiple_input_temp));

    int n = stoi(first_multiple_input[0]);

    int d = stoi(first_multiple_input[1]);

    string expenditure_temp_temp;
    getline(cin, expenditure_temp_temp);

    vector<string> expenditure_temp = split(rtrim(expenditure_temp_temp));

    vector<int> expenditure(n);

    for (int i = 0; i < n; i++) {
        int expenditure_item = stoi(expenditure_temp[i]);

        expenditure[i] = expenditure_item;
    }

    int result = activityNotifications(expenditure, d);

    fout << result << "\n";

    fout.close();

    return 0;
}

string ltrim(const string &str) {
    string s(str);

    s.erase(
        s.begin(),
        find_if(s.begin(), s.end(), not1(ptr_fun<int, int>(isspace)))
    );

    return s;
}

string rtrim(const string &str) {
    string s(str);

    s.erase(
        find_if(s.rbegin(), s.rend(), not1(ptr_fun<int, int>(isspace))).base(),
        s.end()
    );

    return s;
}

vector<string> split(const string &str) {
    vector<string> tokens;

    string::size_type start = 0;
    string::size_type end = 0;

    while ((end = str.find(" ", start)) != string::npos) {
        tokens.push_back(str.substr(start, end - start));

        start = end + 1;
    }

    tokens.push_back(str.substr(start));

    return tokens;
}

File: Fraudulent-Activity-Notifications.java
import java.io.*;
import java.math.*;
import java.security.*;
import java.text.*;
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;
import java.util.regex.*;
import java.util.stream.*;
import static java.util.stream.Collectors.joining;
import static java.util.stream.Collectors.toList;

class Result {

    /*
     * Complete the 'activityNotifications' function below.
     *
     * The function is expected to return an INTEGER.
     * The function accepts following parameters:
     *  1. INTEGER_ARRAY expenditure
     *  2. INTEGER d
     */

    public static void insertInOrder(List<Integer> arr, int x) {
        int pos = Collections.binarySearch(arr, x);
        if (pos < 0) {
            pos = -(pos + 1);
        }
        arr.add(pos, x);
    }

    public static void removeFromOrdered(List<Integer> arr, int x) {
        int pos = Collections.binarySearch(arr, x);
        if (pos >= 0) {
            arr.remove(pos);
        }
    }

    public static int activityNotifications(List<Integer> expenditure, int d) {
        boolean even = (d % 2 == 0);
        int notifications = 0;

        List<Integer> ordered = new ArrayList<>(expenditure.subList(0, d));
        Collections.sort(ordered);

        for (int i = d; i < expenditure.size(); i++) {
            double median;
            if (even) {
                median = (ordered.get(d / 2 - 1) + ordered.get(d / 2)) / 2.0;
            } else {
                median = ordered.get(d / 2);
            }

            if (2 * median <= expenditure.get(i)) {
                notifications++;
            }

            removeFromOrdered(ordered, expenditure.get(i - d));
            insertInOrder(ordered, expenditure.get(i));
        }

        return notifications;
    }

}

public class Solution {
    public static void main(String[] args) throws IOException {
        BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bufferedWriter = new BufferedWriter(new FileWriter(System.getenv("OUTPUT_PATH")));

        String[] firstMultipleInput = bufferedReader.readLine().replaceAll("\\s+$", "").split(" ");

        int n = Integer.parseInt(firstMultipleInput[0]);

        int d = Integer.parseInt(firstMultipleInput[1]);

        List<Integer> expenditure = Stream.of(bufferedReader.readLine().replaceAll("\\s+$", "").split(" "))
            .map(Integer::parseInt)
            .collect(toList());

        int result = Result.activityNotifications(expenditure, d);

        bufferedWriter.write(String.valueOf(result));
        bufferedWriter.newLine();

        bufferedReader.close();
        bufferedWriter.close();
    }
}
